### PR TITLE
[Perf] resolving prolonged `cudastreamsynchronize` execution in z image processing

### PR DESCRIPTION
**Pull request by erfgss**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
There are numerous `cudaStreamSynchronize` operations between different forward passes of the z-image.
cudasync

## Test Plan
The reason for the prolonged cudaStreamSynchronize between different denoise steps lies within the denoising loop in pipeline_z_image.py:
###  Executed at each step
```python
t_norm = timestep[0].item()
```
When calling .item() on a scalar tensor on the GPU, PyTorch must wait for the current CUDA stream to complete execution and copy the scalar back to the CPU.

###  Solution
- **pipeline_z_image.py:** (1) Precompute normalized timesteps before the denoising loop and use a list in the loop instead of `timestep[0].item()` each step, removing per-step GPU→CPU sync. (2) In CFG renormalization, replace the Python check `if new_pos_norm > max_new_norm` with `torch.where(...)` so the branch is on GPU and no scalar sync is triggered.
- **z_image_transformer.py:** Replace padding assignments `x[mask] = x_pad_token` and `cap_feats[mask] = cap_pad_token` (which use aten::index_put_/nonzero and cause sync) with `torch.where(mask_exp, pad_token, x)` so padding is applied with element-wise ops and avoids long cudaStreamSynchronize.
## Test Result
withoutcudasync

##  e2e latency comparison and accuracy comparison
```python
DIFFUSION_ATTENTION_BACKEND=FLASH_ATTN
python text_to_image.py --prompt "a cup of coffee" --height 1024 --width 1024
```
### before
```bash
Adding requests:   0%|                                                                                                                                          | 0/1 [00:00<?, ?it/s]
[Stage-0] INFO 02-02 01:31:00 [manager.py:538] Deactivating all adapters: 0 layers                          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 unit/s, output: 0.00 unit/s]
[Stage-0] WARNING 02-02 01:31:00 [kv_transfer_manager.py:356] No connector available for receiving KV cache
[Stage-0] INFO 02-02 01:31:13 [diffusion_engine.py:80] Generation completed successfully.
[Stage-0] INFO 02-02 01:31:14 [diffusion_engine.py:98] Post-processing completed in 1.1945 seconds
INFO 02-02 01:31:14 [log_utils.py:550] {'type': 'request_level_metrics',
INFO 02-02 01:31:14 [log_utils.py:550]  'request_id': '0_719d649d-64be-4399-a865-e7a2824faae7',
INFO 02-02 01:31:14 [log_utils.py:550]  'e2e_time_ms': 14679.594039916992,
INFO 02-02 01:31:14 [log_utils.py:550]  'e2e_tpt': 0.0,
INFO 02-02 01:31:14 [log_utils.py:550]  'e2e_total_tokens': 0,
INFO 02-02 01:31:14 [log_utils.py:550]  'transfers_total_time_ms': 0.0,
INFO 02-02 01:31:14 [log_utils.py:550]  'transfers_total_bytes': 0,
INFO 02-02 01:31:14 [log_utils.py:550]  'stages': {0: {'stage_gen_time_ms': 14623.814582824707,
INFO 02-02 01:31:14 [log_utils.py:550]                 'num_tokens_out': 0,
INFO 02-02 01:31:14 [log_utils.py:550]                 'num_tokens_in': 0}}}
Processed prompts: 100%|███████████████████████████████████████████████████████████████████████████| 1/1 [00:14<00:00, 14.68s/img, est. speed stage-0 img/s: 0.00, avg e2e_lat: 0.0ms]
INFO 02-02 01:31:14 [omni.py:860] [Summary] {'e2e_requests': 1,████████████████████████████████████| 1/1 [00:14<00:00, 14.68s/img, est. speed stage-0 img/s: 0.00, avg e2e_lat: 0.0ms]
INFO 02-02 01:31:14 [omni.py:860]  'e2e_total_time_ms': 14682.517290115356,
INFO 02-02 01:31:14 [omni.py:860]  'e2e_sum_time_ms': 14679.594039916992,
INFO 02-02 01:31:14 [omni.py:860]  'e2e_total_tokens': 0,
INFO 02-02 01:31:14 [omni.py:860]  'e2e_avg_time_per_request_ms': 14679.594039916992,
INFO 02-02 01:31:14 [omni.py:860]  'e2e_avg_tokens_per_s': 0.0,
INFO 02-02 01:31:14 [omni.py:860]  'wall_time_ms': 14682.517290115356,
INFO 02-02 01:31:14 [omni.py:860]  'final_stage_id': {'0_719d649d-64be-4399-a865-e7a2824faae7': 0},
INFO 02-02 01:31:14 [omni.py:860]  'stages': [{'stage_id': 0,
INFO 02-02 01:31:14 [omni.py:860]              'requests': 1,
INFO 02-02 01:31:14 [omni.py:860]              'tokens': 0,
INFO 02-02 01:31:14 [omni.py:860]              'total_time_ms': 14680.581092834473,
INFO 02-02 01:31:14 [omni.py:860]              'avg_time_per_request_ms': 14680.581092834473,
INFO 02-02 01:31:14 [omni.py:860]              'avg_tokens_per_s': 0.0}],
INFO 02-02 01:31:14 [omni.py:860]  'transfers': []}
Adding requests:   0%|                                                                                                                                          | 0/1 [00:14<?, ?it/s]
Total generation time: 14.6852 seconds (14685.15 ms)
```

50s

### after
```bash
Adding requests:   0%|                                                                                                                                          | 0/1 [00:00<?, ?it/s]
[Stage-0] INFO 02-02 01:38:05 [manager.py:538] Deactivating all adapters: 0 layers                          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 unit/s, output: 0.00 unit/s]
[Stage-0] WARNING 02-02 01:38:05 [kv_transfer_manager.py:356] No connector available for receiving KV cache
[Stage-0] INFO 02-02 01:38:19 [diffusion_engine.py:80] Generation completed successfully.
[Stage-0] INFO 02-02 01:38:20 [diffusion_engine.py:98] Post-processing completed in 1.1712 seconds
INFO 02-02 01:38:20 [log_utils.py:550] {'type': 'request_level_metrics',
INFO 02-02 01:38:20 [log_utils.py:550]  'request_id': '0_112172de-616a-4ce4-b023-6eddbdeff466',
INFO 02-02 01:38:20 [log_utils.py:550]  'e2e_time_ms': 14562.788248062134,
INFO 02-02 01:38:20 [log_utils.py:550]  'e2e_tpt': 0.0,
INFO 02-02 01:38:20 [log_utils.py:550]  'e2e_total_tokens': 0,
INFO 02-02 01:38:20 [log_utils.py:550]  'transfers_total_time_ms': 0.0,
INFO 02-02 01:38:20 [log_utils.py:550]  'transfers_total_bytes': 0,
INFO 02-02 01:38:20 [log_utils.py:550]  'stages': {0: {'stage_gen_time_ms': 14509.912490844727,
INFO 02-02 01:38:20 [log_utils.py:550]                 'num_tokens_out': 0,
INFO 02-02 01:38:20 [log_utils.py:550]                 'num_tokens_in': 0}}}
Processed prompts: 100%|███████████████████████████████████████████████████████████████████████████| 1/1 [00:14<00:00, 14.56s/img, est. speed stage-0 img/s: 0.00, avg e2e_lat: 0.0ms]
INFO 02-02 01:38:20 [omni.py:860] [Summary] {'e2e_requests': 1,████████████████████████████████████| 1/1 [00:14<00:00, 14.56s/img, est. speed stage-0 img/s: 0.00, avg e2e_lat: 0.0ms]
INFO 02-02 01:38:20 [omni.py:860]  'e2e_total_time_ms': 14565.569877624512,
INFO 02-02 01:38:20 [omni.py:860]  'e2e_sum_time_ms': 14562.788248062134,
INFO 02-02 01:38:20 [omni.py:860]  'e2e_total_tokens': 0,
INFO 02-02 01:38:20 [omni.py:860]  'e2e_avg_time_per_request_ms': 14562.788248062134,
INFO 02-02 01:38:20 [omni.py:860]  'e2e_avg_tokens_per_s': 0.0,
INFO 02-02 01:38:20 [omni.py:860]  'wall_time_ms': 14565.569877624512,
INFO 02-02 01:38:20 [omni.py:860]  'final_stage_id': {'0_112172de-616a-4ce4-b023-6eddbdeff466': 0},
INFO 02-02 01:38:20 [omni.py:860]  'stages': [{'stage_id': 0,
INFO 02-02 01:38:20 [omni.py:860]              'requests': 1,
INFO 02-02 01:38:20 [omni.py:860]              'tokens': 0,
INFO 02-02 01:38:20 [omni.py:860]              'total_time_ms': 14563.469886779785,
INFO 02-02 01:38:20 [omni.py:860]              'avg_time_per_request_ms': 14563.469886779785,
INFO 02-02 01:38:20 [omni.py:860]              'avg_tokens_per_s': 0.0}],
INFO 02-02 01:38:20 [omni.py:860]  'transfers': []}
Adding requests:   0%|                                                                                                                                          | 0/1 [00:14<?, ?it/s]
Total generation time: 14.5677 seconds (14567.65 ms)
```
50stepss

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [x] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by erfgss** [comment]
@hsliuustc0106 @david6666666 ptal thx

**Comment by hsliuustc0106** [comment]
please provide e2e latency comparison and accuracy comparison

**Comment by ZJY0516** [comment]
Thank you for your contribution. However, the current profiler output only shows the Python stack and a CPU timeline. We are more concerned with the GPU execution timeline.

**Comment by ZJY0516** [comment]
I tested on H20

```
python diffusion_benchmark_serving.py --port 8000 --model Tongyi-MAI/Z-Image-Turbo --task t2i --dataset random --num-prompts 5 --num-inference-steps 9
```

This pr
```
================= Serving Benchmark Result =================
Model:                                   Tongyi-MAI/Z-Image-Turbo
Dataset:                                 random         
Task:                                    t2i            
--------------------------------------------------
Benchmark duration (s):                  21.21          
Request rate:                            inf            
Max request concurrency:                 1              
Successful requests:                     5/5              
--------------------------------------------------
Request throughput (req/s):              0.24           
Latency Mean (s):                        4.2411         
Latency Median (s):                      4.2104         
Latency P99 (s):                         4.2919         

============================================================
```

main
```
================= Serving Benchmark Result =================
Model:                                   Tongyi-MAI/Z-Image-Turbo
Dataset:                                 random         
Task:                                    t2i            
--------------------------------------------------
Benchmark duration (s):                  21.60          
Request rate:                            inf            
Max request concurrency:                 1              
Successful requests:                     5/5              
--------------------------------------------------
Request throughput (req/s):              0.23           
Latency Mean (s):                        4.3191         
Latency Median (s):                      4.2989         
Latency P99 (s):                         4.4440         

============================================================
```

**Review by ZJY0516 — APPROVED** [review]

**Comment by david6666666** [comment]
should add accuracy comparison

**Comment by erfgss** [comment]
> should add accuracy comparison

I have added it

**Comment by david6666666** [comment]
LGTM
